File: ios/Sources/KeyboardExtension/ConversionContextModel.swift
import Foundation

struct ConversionContextMetadata: Equatable {
    let formatVersion: Int
    let modelVersion: Int
    let unigramCount: Int
    let bigramCount: Int
    let sourceSha256: String

    var entryCount: Int { unigramCount + bigramCount }
}

/// Immutable generated costs decoded from the same FKCTX bytes used by Android.
struct ConversionContextModel {
    private struct SurfaceKey: Hashable { let scalars: [UInt32] }
    private struct BigramKey: Hashable {
        let previous: SurfaceKey
        let next: SurfaceKey
    }

    let metadata: ConversionContextMetadata
    private let unigramCosts: [SurfaceKey: Int]
    private let bigramCosts: [BigramKey: Int]

    static let empty = ConversionContextModel(
        metadata: ConversionContextMetadata(
            formatVersion: 0,
            modelVersion: 0,
            unigramCount: 0,
            bigramCount: 0,
            sourceSha256: String(repeating: "0", count: 64)
        ),
        unigramCosts: [:],
        bigramCosts: [:]
    )

    /// Exact surface bigram first; an unknown pair backs off to next-surface unigram.
    func cost(previousSurface: String?, nextSurface: String) -> Int {
        let next = SurfaceKey(scalars: ConversionText.scalarValues(nextSurface))
        if let previousSurface {
            let previous = SurfaceKey(scalars: ConversionText.scalarValues(previousSurface))
            if let exact = bigramCosts[BigramKey(previous: previous, next: next)] {
                return exact
            }
        }
        return unigramCosts[next] ?? 0
    }

    func hasBigram(previousSurface: String, nextSurface: String) -> Bool {
        bigramCosts[BigramKey(
            previous: SurfaceKey(scalars: ConversionText.scalarValues(previousSurface)),
            next: SurfaceKey(scalars: ConversionText.scalarValues(nextSurface))
        )] != nil
    }

    init(data: Data) throws {
        var reader = ContextModelReader(data: data)
        guard try reader.readBytes(count: 8) == Array("FKCTX001".utf8) else {
            throw ConversionContextModelError.invalidMagic
        }
        let formatVersion = try reader.readUInt16()
        guard formatVersion == 1 else {
            throw ConversionContextModelError.unsupportedFormat(formatVersion)
        }
        let modelVersion = try reader.readUInt16()
        guard modelVersion > 0 else { throw ConversionContextModelError.invalidVersion }
        let unigramCount = try reader.readCount()
        let bigramCount = try reader.readCount()
        guard unigramCount <= 1_000_000,
              bigramCount <= 1_000_000,
              unigramCount + bigramCount <= 1_000_000 else {
            throw ConversionContextModelError.tooManyEntries
        }
        let sourceSha256 = try reader.readBytes(count: 32)
            .map { String(format: "%02x", $0) }
            .joined()

        var unigrams: [SurfaceKey: Int] = [:]
        unigrams.reserveCapacity(unigramCount)
        for _ in 0..<unigramCount {
            let key = SurfaceKey(scalars: ConversionText.scalarValues(try reader.readString()))
            guard unigrams.updateValue(try reader.readInt32(), forKey: key) == nil else {
                throw ConversionContextModelError.duplicateEntry
            }
        }
        var bigrams: [BigramKey: Int] = [:]
        bigrams.reserveCapacity(bigramCount)
        for _ in 0..<bigramCount {
            let previous = SurfaceKey(scalars: ConversionText.scalarValues(try reader.readString()))
            let next = SurfaceKey(scalars: ConversionText.scalarValues(try reader.readString()))
            let key = BigramKey(previous: previous, next: next)
            guard bigrams.updateValue(try reader.readInt32(), forKey: key) == nil else {
                throw ConversionContextModelError.duplicateEntry
            }
        }
        guard reader.isAtEnd else { throw ConversionContextModelError.trailingBytes }
        metadata = ConversionContextMetadata(
            formatVersion: formatVersion,
            modelVersion: modelVersion,
            unigramCount: unigramCount,
            bigramCount: bigramCount,
            sourceSha256: sourceSha256
        )
        unigramCosts = unigrams
        bigramCosts = bigrams
    }

    private init(
        metadata: ConversionContextMetadata,
        unigramCosts: [SurfaceKey: Int],
        bigramCosts: [BigramKey: Int]
    ) {
        self.metadata = metadata
        self.unigramCosts = unigramCosts
        self.bigramCosts = bigramCosts
    }
}

enum ConversionContextModelError: Error, Equatable {
    case truncated
    case invalidMagic
    case unsupportedFormat(Int)
    case invalidVersion
    case tooManyEntries
    case invalidString
    case duplicateEntry
    case trailingBytes
}

private struct ContextModelReader {
    private let bytes: [UInt8]
    private var offset = 0

    init(data: Data) { bytes = Array(data) }
    var isAtEnd: Bool { offset == bytes.count }

    mutating func readUInt16() throws -> Int {
        let value = try readBytes(count: 2)
        return Int(value[0]) | Int(value[1]) << 8
    }

    mutating func readCount() throws -> Int {
        let value = try readUInt32()
        guard value <= UInt32(Int32.max) else { throw ConversionContextModelError.tooManyEntries }
        return Int(value)
    }

    mutating func readUInt32() throws -> UInt32 {
        let value = try readBytes(count: 4)
        return UInt32(value[0]) |
            UInt32(value[1]) << 8 |
            UInt32(value[2]) << 16 |
            UInt32(value[3]) << 24
    }

    mutating func readInt32() throws -> Int {
        Int(Int32(bitPattern: try readUInt32()))
    }

    mutating func readString() throws -> String {
        let length = try readUInt16()
        guard length > 0 else { throw ConversionContextModelError.invalidString }
        let encoded = try readBytes(count: length)
        guard let value = String(bytes: encoded, encoding: .utf8) else {
            throw ConversionContextModelError.invalidString
        }
        return value
    }

    mutating func readBytes(count: Int) throws -> [UInt8] {
        guard count >= 0, offset <= bytes.count - count else {
            throw ConversionContextModelError.truncated
        }
        defer { offset += count }
        return Array(bytes[offset..<(offset + count)])
    }
}

File: app/src/main/java/com/example/furiganakeyboard/conversion/ConversionContextModel.kt
package com.example.furiganakeyboard.conversion

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.CodingErrorAction

data class ConversionContextMetadata(
    val formatVersion: Int,
    val modelVersion: Int,
    val unigramCount: Int,
    val bigramCount: Int,
    val sourceSha256: String,
) {
    val entryCount: Int get() = unigramCount + bigramCount
}

/** Immutable, generated context costs shared byte-for-byte with the iOS runtime. */
class ConversionContextModel private constructor(
    val metadata: ConversionContextMetadata,
    private val unigramCosts: Map<String, Int>,
    private val bigramCosts: Map<BigramKey, Int>,
) {
    /** Exact surface bigram first; an unknown pair backs off to the next-surface unigram. */
    fun cost(previousSurface: String?, nextSurface: String): Int {
        if (previousSurface != null) {
            bigramCosts[BigramKey(previousSurface, nextSurface)]?.let { return it }
        }
        return unigramCosts[nextSurface] ?: 0
    }

    fun hasBigram(previousSurface: String, nextSurface: String): Boolean =
        bigramCosts.containsKey(BigramKey(previousSurface, nextSurface))

    private data class BigramKey(val previous: String, val next: String)

    companion object {
        private const val HEADER_BYTES = 52
        private const val FORMAT_VERSION = 1
        private const val MAX_ENTRIES = 1_000_000
        private val MAGIC = "FKCTX001".toByteArray(Charsets.US_ASCII)
        private val EMPTY = ConversionContextModel(
            ConversionContextMetadata(0, 0, 0, 0, "0".repeat(64)),
            emptyMap(),
            emptyMap(),
        )

        fun empty(): ConversionContextModel = EMPTY

        fun decode(bytes: ByteArray): ConversionContextModel {
            require(bytes.size >= HEADER_BYTES) { "context model header is truncated" }
            val reader = Reader(bytes)
            require(reader.readBytes(MAGIC.size).contentEquals(MAGIC)) {
                "invalid context model magic"
            }
            val formatVersion = reader.readUInt16()
            require(formatVersion == FORMAT_VERSION) {
                "unsupported context model format $formatVersion"
            }
            val modelVersion = reader.readUInt16()
            require(modelVersion > 0) { "invalid context model version" }
            val unigramCount = reader.readUInt32()
            val bigramCount = reader.readUInt32()
            require(unigramCount <= MAX_ENTRIES && bigramCount <= MAX_ENTRIES &&
                unigramCount.toLong() + bigramCount <= MAX_ENTRIES
            ) { "context model entry count is too large" }
            val sourceSha256 = reader.readBytes(32).joinToString("") {
                "%02x".format(it.toInt() and 0xff)
            }

            val unigrams = LinkedHashMap<String, Int>(unigramCount)
            repeat(unigramCount) {
                val surface = reader.readString()
                require(unigrams.put(surface, reader.readInt32()) == null) {
                    "duplicate context unigram"
                }
            }
            val bigrams = LinkedHashMap<BigramKey, Int>(bigramCount)
            repeat(bigramCount) {
                val key = BigramKey(reader.readString(), reader.readString())
                require(bigrams.put(key, reader.readInt32()) == null) {
                    "duplicate context bigram"
                }
            }
            require(reader.isAtEnd) { "context model has trailing bytes" }
            return ConversionContextModel(
                ConversionContextMetadata(
                    formatVersion = formatVersion,
                    modelVersion = modelVersion,
                    unigramCount = unigramCount,
                    bigramCount = bigramCount,
                    sourceSha256 = sourceSha256,
                ),
                unigrams.toMap(),
                bigrams.toMap(),
            )
        }

        private class Reader(private val bytes: ByteArray) {
            private var offset = 0
            val isAtEnd: Boolean get() = offset == bytes.size

            fun readUInt16(): Int {
                requireAvailable(2)
                val value = (bytes[offset].toInt() and 0xff) or
                    ((bytes[offset + 1].toInt() and 0xff) shl 8)
                offset += 2
                return value
            }

            fun readUInt32(): Int {
                requireAvailable(4)
                val value = ByteBuffer.wrap(bytes, offset, 4)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .int
                    .toLong() and 0xffff_ffffL
                offset += 4
                require(value <= Int.MAX_VALUE) { "context model count overflows Int" }
                return value.toInt()
            }

            fun readInt32(): Int {
                requireAvailable(4)
                val value = ByteBuffer.wrap(bytes, offset, 4)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .int
                offset += 4
                return value
            }

            fun readString(): String {
                val length = readUInt16()
                require(length > 0) { "empty context model string" }
                val encoded = readBytes(length)
                return Charsets.UTF_8.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                    .decode(ByteBuffer.wrap(encoded))
                    .toString()
            }

            fun readBytes(count: Int): ByteArray {
                requireAvailable(count)
                return bytes.copyOfRange(offset, offset + count).also { offset += count }
            }

            private fun requireAvailable(count: Int) {
                require(count >= 0 && offset <= bytes.size - count) {
                    "context model is truncated"
                }
            }
        }
    }
}
